subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskDependency.java

```diff
@@ -190,4 +193,103 @@ private void addValue(Object dependency) {
         }
         getMutableValues().add(dependency);
     }
+
+    private static class TaskDependencySet implements Set<Object> {
+        private final Set<Object> delegate = Sets.newHashSet();
+
+        @Override
+        public int size() {
+            return delegate.size();
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return delegate.isEmpty();
+        }
+
+        @Override
+        public boolean contains(Object o) {
+            return delegate.contains(o);
+        }
+
+        @Override
+        public Iterator<Object> iterator() {
+            return delegate.iterator();
+        }
+
+        @Override
+        public Object[] toArray() {
+            return delegate.toArray();
+        }
+
+        @Override
+        public <T> T[] toArray(T[] a) {
+            return delegate.toArray(a);
+        }
+
+        @Override
+        public boolean add(Object o) {
+            return delegate.add(o);
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            if (delegate.remove(o)) {
+                return true;
+            }
+
+            for (Iterator<Object> it = delegate.iterator(); it.hasNext();) {
+                Object obj = it.next();
+                if (obj instanceof ProviderInternal) {
+                    if (((ProviderInternal) obj).getType().isInstance(o)) {
```

Comment:
If I understand this correctly, this could realize all `Provider` task dependencies that have the same type as the task being removed.  I wonder if we could track whether a `TaskProvider` had been realized or not and check before we call `get()` here.  It doesn't seem like there's any way you could have an instantiated task object to remove if the provider had not been realized, so removing a task that is provided by an un-realized provider does not seem to be a use case.

Reply:
I guess, here you would say to compare the name and type if it's a `TaskProvider` without getting the value, we could ignore the case where it's not a `TaskProvider`